src/charm.py

```diff
@@ -3,162 +3,272 @@
+
+    def _update_profiles_container(self, event) -> None:
+        """Updates the Profiles Pebble configuration layer if changed."""
+        if not self.profiles_container.can_connect():
+            self.unit.status = WaitingStatus("Waiting to connect to Profiles container")
+            event.defer()
+            return
+
+        try:
+            self._check_leader()
+            self.unit.status = MaintenanceStatus("Configuring Profiles layer")
+            self._update_profiles_layer()
+        except ErrorWithStatus as error:
+            raise error
+        self.unit.status = MaintenanceStatus("Profiles layer configured")
+
+    def _update_kfam_container(self, event) -> None:
+        """Updates the kfam Pebble configuration layer if changed."""
+        if not self.profiles_container.can_connect():
+            self.unit.status = WaitingStatus("Waiting to connect to kfam container")
+            event.defer()
+            return
+
+        try:
+            self._check_leader()
+            self.unit.status = MaintenanceStatus("Configuring kfam layer")
+            update_layer(
+                self._kfam_container_name,
+                self.kfam_container,
+                self._kfam_pebble_layer,
+                self.log,
+            )
+        except ErrorWithStatus as error:
+            raise error
+        self.unit.status = MaintenanceStatus("kfam layer configured")
+
+    def _on_kubeflow_profiles_ready(self, event):
+        """Define and start a workload using the Pebble API.
+        Learn more about Pebble layers at https://github.com/canonical/pebble
+        """
+        try:
+            self._update_profiles_container(event)
+
+        except ErrorWithStatus as e:
+            self.model.unit.status = e.status
+            if isinstance(e.status, BlockedStatus):
+                self.log.error(str(e.msg))
+            else:
+                self.log.info(str(e.msg))
+
+    def _on_kfam_ready(self, event):
+        """Define and start a workload using the Pebble API.
+        Learn more about Pebble layers at https://github.com/canonical/pebble
+        """
+        try:
+            self._update_kfam_container(event)
+
+        except ErrorWithStatus as e:
+            self.model.unit.status = e.status
+            if isinstance(e.status, BlockedStatus):
+                self.log.error(str(e.msg))
+            else:
+                self.log.info(str(e.msg))
+
+    def _on_remove(self, event):
+        self.unit.status = MaintenanceStatus("Removing k8s resources")
+        manifests = self.k8s_resource_handler.render_manifests()
+        try:
+            delete_many(self.k8s_resource_handler.lightkube_client, manifests)
+        except ApiError as e:
+            self.log.warning(f"Failed to delete resources: {manifests} with: {e}")
+            raise e
+        self.unit.status = MaintenanceStatus("K8s resources removed")
 
     def _send_info(self, event, interfaces):
```

Comment:
Add doc string. This code is not touched, but we should be bringing it up to standard.